utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2112 @@
+      std::bind(&BlobDBImpl::deleteObsFiles, this, std::placeholders::_1));
+  tqueue_.add(bdb_options_.sanity_check_period,
+              std::bind(&BlobDBImpl::sanityCheck, this, std::placeholders::_1));
+  tqueue_.add(bdb_options_.wa_stats_period,
+              std::bind(&BlobDBImpl::waStats, this, std::placeholders::_1));
+  tqueue_.add(bdb_options_.fsync_files_period,
+              std::bind(&BlobDBImpl::fsyncFiles, this, std::placeholders::_1));
+  tqueue_.add(
+      bdb_options_.check_seqf_period,
+      std::bind(&BlobDBImpl::checkSeqFiles, this, std::placeholders::_1));
+}
+
+void BlobDBImpl::shutdown() { shutdown_.store(true); }
+
+void BlobDBImpl::OnFlushBeginHandler(DB* db, const FlushJobInfo& info) {
+  if (shutdown_.load()) return;
+
+  // a callback that happens too soon needs to be ignored
+  if (!db_) return;
+
+  fsyncFiles(false);
+}
+
+Status BlobDBImpl::getAllLogFiles(
+    std::set<std::pair<uint64_t, std::string>>* file_nums) {
+  std::vector<std::string> all_files;
+  Status status = myenv_->GetChildren(blob_dir_, &all_files);
+  if (!status.ok()) {
+    return status;
+  }
+
+  for (const auto& f : all_files) {
+    uint64_t number;
+    FileType type;
+    bool psucc = ParseFileName(f, &number, &type);
+    if (psucc && type == kBlobFile) {
+      file_nums->insert(std::make_pair(number, f));
+    } else {
+      Log(InfoLogLevel::WARN_LEVEL, db_options_.info_log,
+          "Skipping file in blob directory %s parse: %d type: %d", f.c_str(),
+          psucc, ((psucc) ? type : -1));
+    }
+  }
+
+  return status;
+}
+
+Status BlobDBImpl::openAllFiles() {
+  WriteLock wl(&mutex_);
+
+  std::set<std::pair<uint64_t, std::string>> file_nums;
+  Status status = getAllLogFiles(&file_nums);
+
+  if (!status.ok()) {
+    Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+        "Failed to collect files from blob dir: %s status: '%s'",
+        blob_dir_.c_str(), status.ToString().c_str());
+    return status;
+  }
+
+  Log(InfoLogLevel::INFO_LEVEL, db_options_.info_log,
+      "BlobDir files path: %s count: %d min: %d max: %d", blob_dir_.c_str(),
+      file_nums.size(), (file_nums.empty()) ? -1 : (file_nums.begin())->first,
+      (file_nums.empty()) ? -1 : (file_nums.end())->first);
+
+  if (!file_nums.empty())
+    next_file_number_.store((file_nums.rbegin())->first + 1);
+
+  for (auto f_iter : file_nums) {
+    std::string bfpath = BlobFileName(blob_dir_, f_iter.first);
+    uint64_t size_bytes;
+    Status s1 = myenv_->GetFileSize(bfpath, &size_bytes);
+    if (!s1.ok()) {
+      Log(InfoLogLevel::WARN_LEVEL, db_options_.info_log,
+          "Unable to get size of %s. File skipped from open status: '%s'",
+          bfpath.c_str(), s1.ToString().c_str());
+      continue;
+    }
+
+    if (debug_level_ >= 1)
+      Log(InfoLogLevel::WARN_LEVEL, db_options_.info_log,
+          "Blob File open: %s size: %d", bfpath.c_str(), size_bytes);
+
+    std::shared_ptr<BlobFile> bfptr =
```

Comment:
The existing pattern is to use unique_ptr to hold file objects.  shared_ptrs are not free. The above lock free queue attempts to get gains on aligned objects and excessive use of shared_ptr would eat it all up. There is not a real need for shared ownership in many cases though I may certainly miss some.